### DISCUSSION TITLE
Getting OpenGL to work

### DISCUSSION THREAD

**Discussion by IoIxD** [autc04/Retro68#256] [Q&A] [answered]
I'd really like to get Apple's OpenGL SDK to work, but it seems like I can't actually link against the SDK that Apple provided. They seem to be CodeWarrior files? Which I can't link against.

Because the gl functions are expected to be present on the system, I imagine that there's something I need to modify in this compiler (or another codebase related to this) to get it working. But I have no idea where I'd even get started. Can anyone give me a rundown of what I'd need to do?

**Comment by jduerstock** [comment]
If you're referring to linking against the '*.lib' files from the OpenGL SDK, it might be easiest to just rebuild them from the included source.  But perhaps I'm misreading the contents of the SDK.

**Comment by IoIxD** [comment]
Ok so I actually did find out how to compile the sources in a way that lines up with the libraries that Retro68 wants.

But it links against it and then doesn't see the function?

CMake
```cmake
if(PLATFORM MATCHES retroppc)
    add_compile_definitions(FOR_PPC)
    set_target_properties(os9fetch PROPERTIES COMPILE_FLAGS "-ffunction-sections -mcpu=601 -O3 -Wall -Wextra -Wno-unused-parameter")
    set_target_properties(os9fetch PROPERTIES LINK_FLAGS "-Wl,-gc-sections")
    target_link_libraries( os9fetch PowerMgrLib)
    target_link_libraries( os9fetch NameRegistryLib)
    target_link_libraries( os9fetch OpenGL)
endif()
```

Confirming the library has the function and then seeing an error.
```sh
➜  os9fetch git:(master) ✗ powerpc-apple-macos-nm /opt/Retro68/toolchain/powerpc-apple-macos/lib/libOpenGL.a | grep "glGetString"
         U .glGetString
➜  os9fetch git:(master) ✗ ./build.sh
...
[  5%] Linking CXX executable os9fetch.code.bin
/opt/Retro68/toolchain/lib/gcc/powerpc-apple-macos/12.2.0/../../../../powerpc-apple-macos/bin/ld: CMakeFiles/os9fetch.dir/src/detection/gpu.cpp.obj:/home/gavin/Projects/os9fetch/src/detection/gpu.cpp:(.pr+0x14): undefined reference to `.glGetString'
/opt/Retro68/toolchain/lib/gcc/powerpc-apple-macos/12.2.0/../../../../powerpc-apple-macos/bin/ld: CMakeFiles/os9fetch.dir/src/detection/gpu.cpp.obj:/home/gavin/Projects/os9fetch/src/detection/gpu.cpp:(.pr+0x24): undefined reference to `.glGetString'
collect2: error: ld returned 1 exit status
make[2]: *** [CMakeFiles/os9fetch.dir/build.make:353: os9fetch.xcoff] Error 1
make[1]: *** [CMakeFiles/Makefile2:85: CMakeFiles/os9fetch.dir/all] Error 2
make: *** [Makefile:91: all] Error 2
```

**Comment by IoIxD** [comment]
Oh wait, I should learn how nm works. The output of that says it's undefined.

So I'm back to square one.

**Comment by jduerstock** [comment]
Try linking against OpenGLLibraryStub, possibly?

$ DumpPEF OpenGLLibraryStub | grep -E "glGetString|Export"
0084 nbrExportSyms    = 0x00000198 (408)
     ------ Loader Export Symbol Table ------
           Section    Address       Class     [Hashword]    Export Name
250A [223] absolute   00000000     TVector    [000B232E]    glGetString

**Reply by IoIxD** [reply]
How?

**Reply by jduerstock** [reply]
target_link_libraries(os9fetch OpenGLLibaryStub)

possibly?

I am admittedly unfamiliar with Retro68 and CMake, but I am interested in seeing this work.

**Reply by autc04** [reply]
OpenGLLibraryStub is a library stub in PEF format, the GNU Linker cannot use that directly.
You should be able to use the MakeImport tool that is part of Retro68 to convert it/generate a stub library in XCOFF format.

**Reply by IoIxD** [reply]
> OpenGLLibraryStub is a library stub in PEF format, the GNU Linker cannot use that directly. You should be able to use the MakeImport tool that is part of Retro68 to convert it/generate a stub library in XCOFF format.

```zsh
➜  Libraries MakeImport OpenGLUtilityStub OpenGL.a
No 'cfrg' resource found.
➜  Libraries 
```

Seems I need to run it on a Mac. I do not have a modern Mac capable of doing this; I tried it over AppleTalk/mount_afp but it does not seem to make a difference.

**Reply by IoIxD** [reply]
If anyone does have these tools and could convert them for me, the files are avaliable here:

https://staticky.com/dl/ftp.apple.com/developer/opengl/SDK/

**Comment by IoIxD** [comment, marked as answer]
I was finally able to get my hands on a modern Mac, and using that, I was able to convert not only the OpenGL libraries, ~~but also SDL 1.2 as a bonus!~~

They are avaliable here: https://github.com/IoIxD/OpenGLOnMacOS9

Update 2025: SDL 1.2 was broken in that repo, and this was updated to just be about OpenGL

**Reply by autc04** [reply]
Sorry I couldn't assist earlier,.
Note that you do *not* need a modern Mac for the conversion, all you need is a way to extract the SDK package into a resource-fork preserving file format (such as MacBinary or AppleDouble).

But I didn't have ready-made instructions on how to do that until I had the time to come up with a working script.

**Comment by autc04** [comment]
For the Nix users, there is now a sample program with complete build scripts at https://github.com/autc04/RetroOpenGL/ .
